src/stringProcessing/relevantWords.js

```diff
@@ -1,312 +1,188 @@
- * @param {number} upperLimit The upper limit of keyword density.
- * @returns {WordCombination[]} Filtered word combinations.
- */
-function filterOnDensity( wordCombinations, wordCount, lowerLimit, upperLimit ) {
-	return wordCombinations.filter( function( combination ) {
-		return ( combination.getDensity( wordCount ) >= lowerLimit && combination.getDensity( wordCount ) < upperLimit
+	for ( let i = 1; i < wordCombinations.length; i++ ) {
+		const currentWord = new WordCombination(
+			wordCombinations[ i ].getWord(),
+			wordCombinations[ i ].getStem(),
+			wordCombinations[ i ].getOccurrences()
 		);
-	} );
-}
 
-/**
- * Filters combinations based on whether they end with a specified string or not.
- *
- * @param {WordCombination[]} wordCombinations The array of WordCombinations to filter.
- * @param {string} str The string the WordCombinations that need to be filtered out end with.
- * @param {string[]} exceptions The array of strings containing exceptions to not filter.
- * @returns {WordCombination[]} The filtered array of WordCombinations.
- */
-function filterEndingWith( wordCombinations, str, exceptions ) {
-	wordCombinations = wordCombinations.filter( function( combination ) {
-		const combinationstr = combination.getCombination();
-		for ( let i = 0; i < exceptions.length; i++ ) {
-			if ( combinationstr.endsWith( exceptions[ i ] ) ) {
-				return true;
+		/*
+		Compare the stem of the current word in the loop with the previously available stem. If they equal, word combinations should be collapsed.
+		When collapsing, the numbers of occurrences get summed.
+		If the stem happens to equal the real word that occured in the text, we can be sure it's ok to display it to the customer.
+		So, the stem reassigns the word.
+		 */
+		if ( currentWord.getStem() ===  previousWord.getStem() ) {
+			previousWord.setOccurrences( previousWord.getOccurrences() + currentWord.getOccurrences() );
+
+			if ( currentWord.getWord() === previousWord.getStem() ) {
+				previousWord.setWord( currentWord.getWord() );
 			}
+		} else {
+			collapsedRelevantWords.push( previousWord );
+			previousWord = currentWord;
 		}
-		return ! combinationstr.endsWith( str );
-	} );
-	return wordCombinations;
+	}
+
+	collapsedRelevantWords.push( previousWord );
+
+	return collapsedRelevantWords;
 }
 
 /**
- * Filters the list of word combination objects based on the language-specific function word filters.
- * Word combinations with specific parts of speech are removed.
+ * Retrieves a function words list from the factory. Returns an empty array if the language does not have function words.
+ *
+ * @param {string} language The language to retrieve function words for.
  *
- * @param {Array} combinations The list of word combination objects.
- * @param {Function} functionWords The function containing the lists of function words.
- * @returns {Array} The filtered list of word combination objects.
+ * @returns {Array} A list of function words for the language.
```

Comment:
Hmm, I'm not 100% sure, but it's just strings in that array right?

```suggestion
 * @returns {string[]} A list of function words for the language.
```

Reply:
It is.